src/vmm/src/devices/virtio/iovec.rs

```diff
@@ -205,15 +208,22 @@ impl IoVecBuffer {
     }
 }
 
+#[derive(Debug)]
+pub struct ParsedDescriptorChain {
+    pub head_index: u16,
+    pub length: u32,
+    pub nr_iovecs: u16,
+}
+
 /// This is essentially a wrapper of a `Vec<libc::iovec>` which can be passed to `libc::readv`.
 ///
 /// It describes a write-only buffer passed to us by the guest that is scattered across multiple
 /// memory regions. Additionally, this wrapper provides methods that allow reading arbitrary ranges
 /// of data from that buffer.
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct IoVecBufferMut {
     // container of the memory regions included in this IO vector
-    vecs: Vec<iovec>,
+    vecs: IovDeque,
     // Total length of the IoVecBufferMut
     len: u32,
```

Comment:
The main reason I left `u32` here is to not have conversions in other parts of the code (especially vsock) where `u32` is the main type. (there was a PR that made all types to be `u32` and this removed a bunch of `try_from().unwrap()`).
Regarding `u32` overflow, we have a check [here](https://github.com/firecracker-microvm/firecracker/pull/4844/files/c35366792c983693de7febfa788cf72c262be48e..e80c2f87837eb88fe0b12f58b1f74d5e4a0b9219#diff-795f655b4265f2269ad7b6300757e79f2190afc77a1377cf8a1eff9a2dae5f63R276) that will prevent this from papenning. Also it does not really matter what type we use. `usize` can overflow as well if guest will try hard enough.

Reply:
> usize can overflow as well if guest will try hard enough.

It can't happen. `256 * u32::MAX` can't overflow a `usize`.

> Regarding u32 overflow, we have a check here that will prevent this from papenning

So, it is true that most likely this won't happen in our case, because we are only using multiple DescriptorChain objects in network devices and these should not give us such big buffers. So, let's keep it like that, but can you please add a comment on top of its definition mentioning why this is ok in the particular case?

Reply:
done